app/javascript/dashboard/routes/dashboard/settings/reports/components/heatmaps/BaseHeatmap.vue

```diff
@@ -126,35 +161,41 @@ function getHeatmapLevelClass(value) {
           :key="ii"
           class="flex items-center justify-center"
         >
-          {{ ii - 1 }} – {{ ii }}
+          {{ ii - 1 }}
         </div>
       </div>
     </template>
     <template v-else>
       <div class="grid gap-[5px] flex-shrink-0">
         <div
-          v-for="dateKey in processedData.keys()"
-          :key="dateKey"
+          v-for="row in dataRows"
+          :key="row.dateKey"
+          v-memo="[row.dateKey]"
           class="h-8 min-w-[70px] text-n-slate-12 text-[10px] font-semibold flex flex-col items-end justify-center"
         >
-          {{ getDayOfTheWeek(new Date(dateKey)) }}
+          {{ getDayOfTheWeek(new Date(row.dateKey)) }}
           <time class="font-normal text-n-slate-11">
-            {{ formatDate(dateKey) }}
+            {{ formatDate(row.dateKey) }}
           </time>
         </div>
       </div>
-      <div class="grid gap-[5px] w-full min-w-[700px]">
+      <div
+        class="grid gap-[5px] w-full min-w-[700px]"
+        style="content-visibility: auto"
+      >
         <div
-          v-for="dateKey in processedData.keys()"
-          :key="dateKey"
+          v-for="row in dataRows"
+          :key="row.dateKey"
+          v-memo="[row.dataHash, colorScheme]"
           class="grid gap-[5px] grid-cols-[repeat(24,_1fr)]"
+          style="content-visibility: auto; contain-intrinsic-size: auto 32px"
```

Comment:
This CSS style helps with perf in two ways

1. The visibility auto will ensure the browser does not needlessly render this (kind of like virtualization)
2. The intrinsic size hint helps the browser calculate scroll heights and some more things helping with perf.